mlir/lib/Dialect/Presburger/Set.cpp

```diff
@@ -0,0 +1,349 @@
+    // we want to reference B.
+    addInequalityFromEquality(eq, false, false);
+    addInequalityFromEquality(eq, true, false);
+  }
+
+  // offset = 2 * S_i.getNumEqualities();
+  for (size_t j = 0; j < S_i.getNumInequalities(); j++) {
+    /*if (isMarkedRedundant[offset + j])
+      continue;*/
+    const auto &ineq = S_i.getInequality(j);
+
+    size_t snap = simplex.getSnapshot();
+
+    SmallVector<int64_t, 64> complement;
+    for (auto coeff : ineq)
+      complement.emplace_back(-coeff);
+
+    // The constant is at the end
+    --complement[ineq.size() - 1];
+
+    B.addInequality(complement);
+    simplex.addInequality(complement);
+    subtractRecursively(B, simplex, S, i + 1, result);
+    B.removeInequality(B.getNumInequalities() - 1);
+    simplex.rollback(snap);
+
+    B.addInequality(ineq);
+    simplex.addInequality(ineq);
+  }
+
+  // TODO benchmark technically we can probably drop this as the caller will
+  // rollback. See if it makes much of a difference. Only the last rollback
+  // would be eliminated by this.
+  simplex.rollback(initialSnap);
+}
+
+// Returns the set difference B - S.
+PresburgerSet PresburgerSet::subtract(FlatAffineConstraints c,
+                                      const PresburgerSet &set) {
+  assert(set.getNumDims() == c.getNumDimIds() &&
+         set.getNumSyms() == c.getNumSymbolIds() &&
+         "Sets to be subtracted have different dimensionality");
+  if (c.isEmptyByGCDTest())
+    return PresburgerSet::makeEmptySet(c.getNumDimIds(), c.getNumSymbolIds());
+
+  PresburgerSet result(c.getNumDimIds());
+  Simplex simplex(c);
+  subtractRecursively(c, simplex, set, 0, result);
+  return result;
+}
+
+PresburgerSet PresburgerSet::complement(const PresburgerSet &set) {
+  return subtract(FlatAffineConstraints(set.getNumDims(), set.getNumSyms()),
+                  set);
+}
+
+// Subtracts the set S from the current set.
+//
+// We compute (U_i T_i) - (U_i S_i) as U_i (T_i - U_i S_i).
+void PresburgerSet::subtract(const PresburgerSet &set) {
+  assert(set.getNumDims() == nDim && set.getNumSyms() == nSym &&
+         "Sets to be subtracted have different dimensionality");
+  if (markedEmpty)
+    return;
+  if (set.isMarkedEmpty())
+    return;
+  if (set.isUniverse()) {
+    markedEmpty = true;
+    return;
+  }
+  if (isUniverse()) {
+    *this = PresburgerSet::complement(set);
+    return;
+  }
+
+  PresburgerSet result = PresburgerSet::makeEmptySet(nDim, nSym);
+  for (const FlatAffineConstraints &c : flatAffineConstraints)
+    result.unionSet(subtract(c, set));
+  *this = result;
+}
+
+bool PresburgerSet::equal(const PresburgerSet &s, const PresburgerSet &t) {
+  // TODO implemenmt this
+  return false;
+}
+
+// TODO refactor and rewrite after discussion with the others
+void PresburgerSet::print(raw_ostream &os) const {
+  printVariableList(os);
+  if (markedEmpty) {
+    // TODO dicuss what we want to print in the empty case
+    // os << ": (1 = 0)";
+    // return;
```

Comment:
In libint you just add '"<marked empty>"' do you suggest that I do the same here? Or is a trivially false constraint better as it can be parsed without additional work. 
Note that this is probably broken in libint as I was not able to find anything with "empty" in the parser.

Reply:
I think a trivially false constraint is fine.

Yes, the libint parser can't parse this at the moment.